geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/WanCopyRegionFunction.java

```diff
@@ -0,0 +1,480 @@
+      if (event != null) {
+        batch.add(event);
+        batchIndex++;
+      }
+    }
+    return batch;
+  }
+
+  Set<?> getEntries(Region region, GatewaySender sender) {
+    if (region instanceof PartitionedRegion && sender.isParallel()) {
+      return ((PartitionedRegion) region).getDataStore().getAllLocalBucketRegions()
+          .stream()
+          .flatMap(br -> ((Set<?>) br.entrySet()).stream()).collect(Collectors.toSet());
+    }
+    return region.entrySet();
+  }
+
+  @VisibleForTesting
+  GatewayQueueEvent createGatewaySenderEvent(InternalCache cache,
+      InternalRegion region, GatewaySender sender, Region.Entry entry) {
+    final EntryEventImpl event;
+    if (region instanceof PartitionedRegion) {
+      event = createEventForPartitionedRegion(sender, cache, region, entry);
+    } else {
+      event = createEventForReplicatedRegion(cache, region, entry);
+    }
+    if (event == null) {
+      return null;
+    }
+    try {
+      return new GatewaySenderEventImpl(EnumListenerEvent.AFTER_UPDATE_WITH_GENERATE_CALLBACKS,
+          event, null, true);
+    } catch (IOException e) {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  final CliFunctionResult cancelWanCopyRegion(FunctionContext<Object[]> context,
+      String regionName, String senderId) {
+    boolean found = false;
+    String threadBaseName = getFunctionThreadName(regionName, senderId);
+    for (Thread t : Thread.getAllStackTraces().keySet()) {
+      if (t.getName().startsWith(threadBaseName)) {
+        found = true;
+        t.interrupt();
+      }
+    }
+    if (found) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          CliStrings.WAN_COPY_REGION__MSG__EXECUTION__CANCELED);
+    }
+    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+        CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__NO__RUNNING__COMMAND,
+            regionName, senderId));
+  }
+
+  public static String getFunctionThreadName(String regionName, String senderId) {
+    return "wanCopyRegionFunctionThread_" + regionName + "_" + senderId;
+  }
+
+  /**
+   * It runs the actions to be done after a batch has been
+   * sent: throw an interrupted exception if the operation was canceled and
+   * adjust the rate of copying by sleeping if necessary.
+   *
+   * @param startTime time at which the entries started to be copied
+   * @param copiedEntries number of entries copied so far
+   * @param maxRate maximum copying rate
+   */
+  void doPostSendBatchActions(long startTime, int copiedEntries, long maxRate)
+      throws InterruptedException {
+    if (Thread.currentThread().isInterrupted()) {
+      throw new InterruptedException();
+    }
+    long sleepMs = getTimeToSleep(startTime, copiedEntries, maxRate);
+    if (sleepMs > 0) {
+      logger.info("{}: Sleeping for {} ms to accommodate to requested maxRate",
+          this.getClass().getName(), sleepMs);
```

Comment:
`getClass().getName()` provides the fully qualified name (FQN) which is full.package.ClassName. `getClass().getSimpleName()` gives you just ClassName. Just making sure you know about that in case you didn't want the package.